blocks/bricks/wrappers.py

```diff
@@ -2,40 +2,91 @@
 from theano import tensor
 
 from blocks.bricks.base import Brick, application
+from blocks.utils import dict_subset, pack
 
 
-class As2D(Brick):
-    """Wraps a 2D application around reshapes to accomodate N-D tensors.
+class WithExtraDims(Brick):
```

Comment:
There are no flaws in your reasoning, and in principle what you suggest is good - have exactly one function do exactly one task. But in this particular situation I feel that wrapping the apply with another brick approach is not good - it is not intuitive and requires to instantiate softmax twice. Also, the usage is more difficult - instead of saying `SoftMax(axis=-1)`, i.e. take softmax along last axis you have to wrap it by saying how many dimensions to add - and separately for each of softmax applys that you want to use. And you have to double wrap it to do e.g. `SoftMax(axis=0)` on 3d tensors because you also need the dimshuffle.

Stretching the unix analogy, `sort` has the option to specify a sort key, even though you can achieve the same chaining a cut to reorder the fields, a simple sort, and another cut to reorder it back. 

An acceptable solution would be to somehow outsource the flattening-reshaping, so that you can still say `SoftMax(axis=-1)`, but you have a single implementation of the dimshuffle-flatten-reshape-dimshuffle logic. But this is a form of refactoring and can wait until we have at least two nearly-identical softmax implementations. Especially, that softmax is strange - it typically takes as inputs tensors with different number of dimensions.